### DISCUSSION TITLE
How do I use environment variables to programatically create the first admin user ?

### DISCUSSION THREAD

**Discussion by CUexter** [pocketbase/pocketbase#1797] [Q&A] [answered]
https://github.com/pocketbase/pocketbase/discussions/1354

I looked at this and just wonder if there is a way to do this using environment variables

I tried this but it doesn't work 
```js
require("dotenv").config();

let email;
let password;

if (process.env.NODE_ENV === "development" || process.env.NODE_ENV === "test") {
  email = process.env.admin_email ?? "admin@example.com";
  password = process.env.admin_pw ?? "admin123456";
} else {
  email = process.env.admin_email;
  password = process.env.admin_pw;
}

if (!email || !password)
  throw new Error("email and password must be defined in production");

migrate(
  (db) => {
    const admin = new Admin();

    admin.email = email;
    admin.setPassword(password);
    return Dao(db).saveAdmin(admin);
  },
  (db) => {
    const dao = new Dao(db);
    const admin = dao.findAdminByEmail(email);
    return dao.deleteAdmin(admin);
  }
);

```

**Comment by ganigeorgiev** [comment, marked as answer]
The embedded JS runtime is not a node environment and things like `process.env` and other node specifics will not work out of the box.

It is possible to expose some of the Go `os` package functions (_or register the `goja_nodejs/process` module_) but I'm not sure that I understand the use case. Could you elaborate a little more why do you need to create an admin in a migration with env variables? In general, I'm not sure if using env variable is a good idea in a migration because the variables can change and in case of a failure and rollback for example it may accidentally delete some other admin rather than the initially created one (_and in this case It may be better to even leave the rollback empty if you really need to rely on some dynamic value_).

**Reply by TechLionDev** [reply]
Here is my migration file:
```js
/// <reference path="../pb_data/types.d.ts" />
migrate((app) => {
    let settings = app.settings()
    settings.meta.appURL = process.env.POCKETBASE_URL
    settings.logs.maxDays = 30
    settings.logs.logAuthId = true
    settings.logs.logIP = true

    app.save(settings)
})
```
despite having a .env file, appURL is always blank. Please help.

**Reply by ganigeorgiev** [reply]
@TechLionDev Please don't hijack old discussions.

In any case, `.env` has no special meaning in PocketBase. See also https://github.com/pocketbase/pocketbase/discussions/3181#discussioncomment-6816114.

If you still have other questions, please open a new Q&A discussion but try to elaborate more clearly on your use case and the desired behavior.

**Comment by CUexter** [comment]
I would like to be able to do e2e testing automatically, which requires setting up the admin user programmatically.


Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by ganigeorgiev** [reply]
I'm still not really sure that I understand why you need it, but I guess it doesn't hurt to allow accessing the `os.Environ()` map within the migrations. I've enabled the `process.env` goja module in the `develop` branch and it will be available with the next v0.13.0 release.

But once again, you may not need it. For e2e tests you can prepare a separate pb_data with your mocks with not only admin but other data too. You can specify a different `pb_data` directory using the `--dir=/path/to/pb_data_test` flag when starting the executable.

**Comment by CUexter** [comment]
I guess you are right, thank you

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
